core/services/workflows/store_db.go

```diff
@@ -0,0 +1,263 @@
+	ID                  uint
+	WorkflowExecutionID string `db:"workflow_execution_id"`
+	Ref                 string
+	Status              string
+	Inputs              []byte
+	OutputErr           string `db:"output_err"`
+	OutputValue         []byte `db:"output_value"`
+}
+
+// `updateStatus` updates the status of the given workflow execution
+func (d *dbStore) updateStatus(ctx context.Context, executionID string, status string) error {
+	sql := `UPDATE workflow_executions SET status = $1 WHERE id = $2`
+	_, err := d.db.ExecContext(ctx, sql, status, executionID)
+	return err
+}
+
+// `updateStep` updates the given step. This will correspond to an insert, or an update
+// depending on whether a step with the ref already exists.
+func (d *dbStore) updateStep(ctx context.Context, stepState *stepState) (executionState, error) {
+	step, err := stateToStep(stepState)
+	if err != nil {
+		return executionState{}, err
+	}
+
+	err = d.upsertSteps(ctx, []workflowStep{step})
+	if err != nil {
+		return executionState{}, err
+	}
+
+	return d.get(ctx, step.WorkflowExecutionID)
+}
+
+// `get` fetches the executionState from the database.
+func (d *dbStore) get(ctx context.Context, executionID string) (executionState, error) {
+	wexSql := `SELECT * FROM workflow_executions WHERE id = $1`
+	wex := &workflowExecution{}
+	err := d.db.GetContext(ctx, wex, wexSql, executionID)
+	if err != nil {
+		return executionState{}, err
+	}
+
+	wsSql := `SELECT * FROM workflow_steps WHERE workflow_execution_id = $1`
+	ws := []workflowStep{}
+	err = d.db.SelectContext(ctx, &ws, wsSql, wex.ID)
+	if err != nil {
+		return executionState{}, err
+	}
+
+	steps := map[string]*stepState{}
+	for _, s := range ws {
+		ss, err := stepToState(s)
+		if err != nil {
+			return executionState{}, err
+		}
+
+		steps[s.Ref] = ss
+	}
+
+	es := executionState{
+		executionID: wex.ID,
+		workflowID:  wex.WorkflowID,
+		status:      wex.Status,
+		steps:       steps,
+	}
+	return es, nil
+}
+
+func stepToState(step workflowStep) (*stepState, error) {
+	var inputs *values.Map
+	if len(step.Inputs) > 0 {
+		vmProto := &valuespb.Map{}
+		err := proto.Unmarshal(step.Inputs, vmProto)
+		if err != nil {
+			return nil, err
+		}
+
+		inputs = values.FromMapValueProto(vmProto)
+	}
+
+	var (
+		outputErr error
+		outputs   values.Value
+	)
+
+	if len(step.OutputErr) != 0 {
+		outputErr = errors.New(step.OutputErr)
+	}
+
+	if len(step.OutputValue) != 0 {
+		vmProto := &valuespb.Map{}
+		err := proto.Unmarshal(step.OutputValue, vmProto)
+		if err != nil {
+			return nil, err
+		}
+
+		outputs = values.FromMapValueProto(vmProto)
+	}
+
+	var so *stepOutput
+	if outputErr != nil || outputs != nil {
+		so = &stepOutput{
+			err:   outputErr,
+			value: outputs,
+		}
+	}
+
+	return &stepState{
+		executionID: step.WorkflowExecutionID,
+		ref:         step.Ref,
+		status:      step.Status,
+		inputs:      inputs,
+		outputs:     so,
+	}, nil
+}
+
+func stateToStep(state *stepState) (workflowStep, error) {
```

Comment:
I'm really not a fan of adding methods to structs in general... Something about being scarred by the experience of OOP :P

To explain my thinking -- in general, I find adding methods like this contributes to coupling and will start to make things harder to change. Since this is only used by the store, it's fine to just have it as a helper here; if it will be used elsewhere (it won't be since `workflowStep` is the struct describing the row record), then there's an argument for adding it to the struct.